waiter/src/waiter/auth/oidc.clj

```diff
@@ -0,0 +1,248 @@
+  "Returns true if OIDC auth is enabled for the service."
+  [allow-oidc-auth-api? allow-oidc-auth-services? {:keys [waiter-api-call?] :as request}]
+  (let [use-oidc-auth-env (get-in request [:waiter-discovery :service-description-template "env" "USE_OIDC_AUTH"])]
+    (or
+      ;; service requests will enable OIDC auth based on env variable or when absent, allow-oidc-auth-services?
+      (and (not waiter-api-call?)
+           (if (some? use-oidc-auth-env) (= "true" use-oidc-auth-env) allow-oidc-auth-services?))
+      ;; waiter api requests will enable OIDC auth based on allow-oidc-auth-api?
+      (and waiter-api-call? allow-oidc-auth-api?))))
+
+(defn supports-redirect?
+  "Returns true when:
+   - either the request is deemed to have come from a browser
+   - or the accept-redirect=yes request header is present in the request."
+  [oidc-authority request]
+  (or (hu/browser-request? request)
+      ;; Accept-Redirect request header "yes" means the user-agent will follow redirects.
+      ;; Accept-Redirect-Auth request header indicates which authorities the user-agent is willing to redirect to and authenticate at.
+      ;; https://tools.ietf.org/id/draft-williams-http-accept-auth-and-redirect-02.html#rfc.section.2
+      (and (= "yes" (get-in request [:headers accept-redirect-header-name]))
+           (let [accept-redirect-auth (get-in request [:headers accept-redirect-auth-header-name])]
+             (or (str/blank? accept-redirect-auth)
+                 (= "." accept-redirect-auth)
+                 (some #(= oidc-authority %) (str/split accept-redirect-auth #" ")))))))
+
+(defn wrap-auth-handler
+  "Wraps the request handler with a handler to trigger OIDC+PKCE authentication."
+  [{:keys [allow-oidc-auth-api? allow-oidc-auth-services? jwt-auth-server oidc-authorize-uri password]} request-handler]
+  (let [oidc-authority (-> oidc-authorize-uri (URI.) (.getAuthority))]
+    (fn oidc-auth-handler [request]
+      (cond
+        (or (auth/request-authenticated? request)
+            (not (oidc-enabled-on-service? allow-oidc-auth-api? allow-oidc-auth-services? request))
+            ;; OIDC auth is no-op when request cannot be redirected
+            (not (supports-redirect? oidc-authority request)))
+        (request-handler request)
+
+        :else
+        (ru/update-response
+          (request-handler request)
+          (make-oidc-auth-response-updater jwt-auth-server password request))))))
+
+(defrecord OidcAuthenticator [allow-oidc-auth-api? allow-oidc-auth-services? oidc-authorize-uri
+                              jwt-auth-server jwt-validator password])
+
+(defn create-oidc-authenticator
+  "Factory function for creating OIDC authenticator middleware"
+  [jwt-auth-server jwt-validator
+   {:keys [allow-oidc-auth-api? allow-oidc-auth-services? oidc-authorize-uri password]
+    :or {allow-oidc-auth-api? false
+         allow-oidc-auth-services? false}}]
+  {:pre [(satisfies? jwt/AuthServer jwt-auth-server)
```

Comment:
If anything goes wrong here, how will the user know what went wrong?

Reply:
This snippet is triggered only once, during Waiter process start-up, and will cause the process to exit. It is a validation of the dependency injection mechanism in place.